### PR TITLE
Add ability to use a different Deno.connect() implementation for ConnectionTCP

### PR DESCRIPTION
**Pull request by ehsan2003**

### CODE REVIEW

File: connection/1_connection_tcp.ts [hunk 1]
```diff
@@ -25,6 +25,10 @@ import { ConnectionUnframed } from "./0_connection.ts";
 
 const L = getLogger("ConnectionTCP");
 
+export interface TcpConnector {
+  connect: typeof Deno.connect;
+}
+
```
Comment (roj1512) [inline connection/1_connection_tcp.ts:31]:
```suggestion
```

File: connection/1_connection_tcp.ts [hunk 2]
```diff
@@ -35,11 +39,13 @@ export class ConnectionTCP extends ConnectionUnframed implements ConnectionUnfra
   #nextResolve: [number, { resolve: () => void; reject: (err: unknown) => void }] | null = null;
   #canRead = false;
   #canWrite = false;
+  #connector: TcpConnector;
```
Comment (roj1512) [inline connection/1_connection_tcp.ts:42]:
```suggestion
connect: typeof Deno["connect"] = Deno.connect;
```

File: connection/1_connection_tcp.ts [hunk 3]
```diff
@@ -35,11 +39,13 @@ export class ConnectionTCP extends ConnectionUnframed implements ConnectionUnfra
...
 
-  constructor(hostname: string, port: number) {
+  constructor(hostname: string, port: number, connector: TcpConnector = Deno) {
     super();
     this.#hostname = hostname;
     this.#port = port;
+    this.#connector = connector;
```
Comment (roj1512) [inline connection/1_connection_tcp.ts:48]:
```suggestion
  constructor(hostname: string, port: number) {
    super();
    this.#hostname = hostname;
    this.#port = port;
```

File: transport/3_transport_provider_tcp.ts [hunk 4]
```diff
@@ -18,17 +18,18 @@
  * along with this program.  If not, see <https://www.gnu.org/licenses/>.
  */
 
-import { ConnectionTCP } from "../connection/1_connection_tcp.ts";
+import { ConnectionTCP, TcpConnector } from "../connection/1_connection_tcp.ts";
```
Comment (roj1512) [inline transport/3_transport_provider_tcp.ts:21]:
```suggestion
import { ConnectionTCP } from "../connection/1_connection_tcp.ts";
```

File: transport/3_transport_provider_tcp.ts [hunk 5]
```diff
@@ -18,17 +18,18 @@
...
 import { TransportAbridged } from "./1_transport_abridged.ts";
 import { getDcId, getDcIps, TransportProvider } from "./2_transport_provider.ts";
 
 export function transportProviderTcp(
-  params?: { ipv6?: boolean; obfuscated?: boolean },
+  params?: { ipv6?: boolean; obfuscated?: boolean; connector?: TcpConnector },
```
Comment (roj1512) [inline transport/3_transport_provider_tcp.ts:26]:
```suggestion
  params?: { ipv6?: boolean; obfuscated?: boolean; connect?: typeof Deno["connect"] },
```

File: transport/3_transport_provider_tcp.ts [hunk 6]
```diff
@@ -18,17 +18,18 @@
...
 ): TransportProvider {
   return ({ dc, cdn }) => {
     const connection = new ConnectionTCP(
       getDcIps(dc, params?.ipv6 ? "ipv6" : "ipv4")[0],
       80,
+      params?.connector,
     );
```
Comment (roj1512) [inline transport/3_transport_provider_tcp.ts:33]:
```suggestion
    const connection = new ConnectionTCP(
      getDcIps(dc, params?.ipv6 ? "ipv6" : "ipv4")[0],
      80,
    );
    if (params?.connect) {
      connection.connect = params.connect;
    }
```

**Review by roj1512 — CHANGES_REQUESTED** [review]
To make this addition eve more straightforward, we can instead have a public property in ConnectionTCP that is assigned to `Deno.connect` by default.

**Comment by ehsan2003** [comment]
> To make this addition eve more straightforward, we can instead have a public property in ConnectionTCP that is assigned to `Deno.connect` by default.

but this requires that class to be mutated right after creation
I believe mutation after starting client isn't meaningful so why not constructor input ? 

it could lead to misconception of the ability to mutate it in runtime which is not the case

**Comment by ehsan2003** [comment]
but we can simply remove that type and make the connector just a simple function to make it simpler

**Comment by roj1512** [comment]
Good points. I think you can do what I suggested, but make `connect` a private property (rename it to `#connect`), and accept its value from the constructor.

File: connection/1_connection_tcp.ts [hunk 7]
```diff
@@ -25,6 +25,7 @@ import { ConnectionUnframed } from "./0_connection.ts";
 
 const L = getLogger("ConnectionTCP");
 
+
```
Comment (roj1512) [inline connection/1_connection_tcp.ts:28]:
```suggestion
```

File: connection/1_connection_tcp.ts [hunk 8]
```diff
@@ -35,11 +36,13 @@ export class ConnectionTCP extends ConnectionUnframed implements ConnectionUnfra
   #nextResolve: [number, { resolve: () => void; reject: (err: unknown) => void }] | null = null;
   #canRead = false;
   #canWrite = false;
+  #connect: typeof Deno.connect;
 
-  constructor(hostname: string, port: number) {
+  constructor(hostname: string, port: number, connect = Deno.connect) {
```
Comment (roj1512) [inline connection/1_connection_tcp.ts:41]:
```suggestion
  constructor(hostname: string, port: number, connect: typeof Deno.connect = Deno.connect) {
```

**Review by roj1512 — APPROVED** [review]
Looks good, thank you so much!

**Pull request merged**